listenbrainz/webserver/views/feedback_api.py

```diff
@@ -0,0 +1,167 @@
+        Max: :data:`~webserver.views.api.MAX_ITEMS_PER_GET`.
+    :type count: ``int``
+    :param offset: Optional, number of feedback to skip from the beginning, for pagination.
+        Ex. An offset of 5 means the top 5 feedback will be skipped, defaults to 0.
+    :type offset: ``int``
+    :statuscode 200: Yay, you have data!
+    :resheader Content-Type: *application/json*
+    """
+
+    score = _parse_int_arg('score')
+
+    offset = _get_non_negative_param('offset', default=0)
+    count = _get_non_negative_param('count', default=DEFAULT_ITEMS_PER_GET)
+
+    count = min(count, MAX_ITEMS_PER_GET)
+
+    user = db_user.get_by_mb_id(user_name)
+    if user is None:
+        raise APINotFound("Cannot find user: %s" % user_name)
+
+    if score is None:
+        feedback = db_feedback.get_feedback_by_user_id(user["id"])
+    else:
+        if score in [-1, 1]:
+            feedback = db_feedback.get_feedback_by_user_id_and_score(user["id"], score)
+        else:
+            log_raise_400("Score can have a value of 1 or -1.", request.args)
+
+    total_count = len(feedback)
+    feedback_list = feedback[offset:(offset+count)]
+
+    for i, fb in enumerate(feedback_list):
+        fb.user_id = user_name
+        feedback_list[i] = fb.dict()
+
+    return jsonify({
+        "feedback": feedback_list,
+        "count": len(feedback_list),
+        "total_count": total_count,
+        "offset": offset
+    })
+
+
+@feedback_api_bp.route("/recording/<recording_msid>/get-feedback", methods=["GET"])
+@crossdomain()
+@ratelimit()
+def get_feedback_for_recording(recording_msid):
+    """
+    Get feedback for recording with given ``recording_msid``. The format for the JSON returned
+    is defined in our :ref:`feedback-json-doc`.
+
+    :param count: Optional, number of feedback to return, Default: :data:`~webserver.views.api.DEFAULT_ITEMS_PER_GET`
+        Max: :data:`~webserver.views.api.MAX_ITEMS_PER_GET`.
+    :type count: ``int``
+    :param offset: Optional, number of feedback to skip from the beginning, for pagination.
+        Ex. An offset of 5 means the top 5 feedback will be skipped, defaults to 0.
+    :type offset: ``int``
+    :statuscode 200: Yay, you have data!
+    :resheader Content-Type: *application/json*
+    """
+
+    if not is_valid_uuid(recording_msid):
+        log_raise_400("%s MSID format invalid." % recording_msid)
+
+    offset = _get_non_negative_param('offset', default=0)
+    count = _get_non_negative_param('count', default=DEFAULT_ITEMS_PER_GET)
+
+    count = min(count, MAX_ITEMS_PER_GET)
+
+    feedback = db_feedback.get_feedback_by_recording_msid(recording_msid)
+
+    total_count = len(feedback)
+    feedback_list = feedback[offset:(offset+count)]
+
+    user_id_list = [fb.user_id for fb in feedback_list]
+    user_name_list = db_user.get_users_in_order(user_id_list)
+
+    for i, fb in enumerate(feedback_list):
+        fb.user_id = user_name_list[i]["musicbrainz_id"]
+        feedback_list[i] = fb.dict()
```

Comment:
When getting data from the database, it's often a good idea to let the database do as much work as possible. Partially so that you don't make a mistake when joining data together, but also because the database is much more efficient at collecting this data and returning if you request it all in one go.

In this case I see a few things that can be improved:
 * You perform the offset / count in python
 * you perform 2 queries - one for the feedback and one for the users
 * you manually link feedback and users together

I would make the following changes:
* In the database query, use OFFSET ... LIMIT to do the paging. You can pass these parameters into the `get_feedback_by_recording_msid` method.
* Directly join the user table against the feedback table (`select ... blah, u.username from recording_feedback join "user" u on u.id = recording_feedback.user_id`), so that you get the feedback and the username in a single query. I think it makes sense to also include username in the `Feedback` model here. You could include both the user id and the username, making the id required (as it's needed for the database row). In some cases the username could be optional (e.g. when you're making an object to pass it to the insert method).

Reply:
Done in #883